### PR TITLE
Files for key vault misconfiguration alert

### PR DESCRIPTION
**Pull request by jaesoni**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [/] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

**Comment by azure-quickstarts** [comment]
@jaesoni - check this PR for updates that may be needed to documentation that references this sample.  [This is an automated message. You are receiving it because you are listed as the docOwner in metadata.json.]

File: demos/ag-alert-keyvault-advisor/prereqs/prereq.azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+        "name": "[variables('actionGroupName')]",
+        "location": "Global",
+        "properties": {
+            "groupShortName": "[variables('actionGroupName')]",
+            "enabled": true
+        }
+    }
+  ],
+  "outputs": {
+        "applicationGatewayResourceId": {
```
Comment (mainak-agrawal) [inline demos/ag-alert-keyvault-advisor/prereqs/prereq.azuredeploy.json:355]: This variable name needs to be the same as the suffix (<GET-PREREQ-SUFFIX>) in the main parameters file

File: demos/ag-alert-keyvault-advisor/azuredeploy.parameters.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
+   "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+   "contentVersion": "1.0.0.0",
+   "parameters": {
+      "alertName": {
+         "value": "Application Gateway alert for misconfigured key vault"
+      },
+      "applicationGatewayResourceGroup": {
+         "value": "GET-PREREQ-applicationGatewayResourceGroup"
```
Comment (mainak-agrawal) [inline demos/ag-alert-keyvault-advisor/azuredeploy.parameters.json:9]: This needs to be corrected as explained in the comment on prereq template file

File: demos/ag-alert-keyvault-advisor/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      },
+      "actionGroupResourceID":{
+         "type":"String",
+         "metadata":{
+            "description":"Action Group ResourceID example /subscriptions/subs-id/resourceGroups/rg-name/providers/microsoft.insights/actiongroups/action-group-name"
+         }
+      }
+   },
+   "variables":{
+
```
Comment (mainak-agrawal) [inline demos/ag-alert-keyvault-advisor/azuredeploy.json:22]: Can avoid empty line here

File: demos/ag-alert-keyvault-advisor/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                     "equals":"Microsoft.Advisor/recommendations/available/action"
+                  }
+               ]
+            },
+            "actions":{
+               "actionGroups":[
+                  {
+                     "actionGroupId":"[parameters('actionGroupResourceID')]",
+                     "webhookProperties":{
+
```
Comment (mainak-agrawal) [inline demos/ag-alert-keyvault-advisor/azuredeploy.json:55]: Can avoid empty line here

File: demos/ag-alert-keyvault-advisor/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+   "$schema":"https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+   "contentVersion":"1.0.0.0",
+   "parameters":{
+      "alertName":{
+         "type":"String"
+      },
+      "applicationGatewayResourceGroup":{
+         "type":"String",
+         "metadata":{
+            "description":"Resource Group example /subscriptions/subs-id/resourceGroups/rg-name"
```
Comment (mainak-agrawal) [inline demos/ag-alert-keyvault-advisor/azuredeploy.json:11]: Resource group's resource ID seems to be required here, but the Prereq is populating Application Gateway resource ID

File: demos/ag-alert-keyvault-advisor/README.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
+# Setup alert for misconfigured Key Vault for your Azure Application Gateway.
+
+![Azure Public Test Date](https://azurequickstartsservice.blob.core.windows.net/badges/demos/ag-alert-backend-5xx/PublicLastTestDate.svg)
```
Comment (mainak-agrawal) [inline demos/ag-alert-keyvault-advisor/README.md:3]: Please update all of these lines with the updated alert name. These are currently using "ag-alert-backend-5xx" in the addresses.

File: demos/ag-alert-keyvault-advisor/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+   },
+   "resources":[
+      {
+         "type":"microsoft.insights/activityLogAlerts",
+         "apiVersion":"2020-10-01",
+         "name":"[parameters('alertName')]",
+         "location":"Global",
+         "properties":{
+            "scopes":[
+               "[parameters('applicationGatewayResourceGroup')]"
```
Comment (mainak-agrawal) [inline demos/ag-alert-keyvault-advisor/azuredeploy.json:32]: From previous published alert, continuing with App Gateway's Resource ID itself should be fine. Don't think we need to change to Resource group ID for this one as the scope identical to the other alert should work.

**Comment by azure-quickstarts** [comment]
@jaesoni - check this PR for updates that may be needed to documentation that references this sample.  [This is an automated message. You are receiving it because you are listed as the docOwner in metadata.json.]

**Comment by azure-quickstarts** [comment]
@jaesoni - check this PR for updates that may be needed to documentation that references this sample.  [This is an automated message. You are receiving it because you are listed as the docOwner in metadata.json.]

File: demos/ag-alert-keyvault-advisor/prereqs/prereq.azuredeploy.json (hunk 1 above)
Comment (jaesoni) [reply demos/ag-alert-keyvault-advisor/prereqs/prereq.azuredeploy.json]: updated

**Comment by azure-quickstarts** [comment]
@jaesoni - check this PR for updates that may be needed to documentation that references this sample.  [This is an automated message. You are receiving it because you are listed as the docOwner in metadata.json.]

File: demos/ag-alert-keyvault-advisor/metadata.json [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{
+  "$schema": "https://aka.ms/azure-quickstart-templates-metadata-schema#",
+  "itemDisplayName": "Create an alert for misconfigured Key Vault on your Azure Application Gateway.",
```
Comment (mainak-agrawal) [inline demos/ag-alert-keyvault-advisor/metadata.json:3]: Max length for this is 60. It is being exceeded. Please update.

**Comment by azure-quickstarts** [comment]
@jaesoni - check this PR for updates that may be needed to documentation that references this sample.  [This is an automated message. You are receiving it because you are listed as the docOwner in metadata.json.]

**Comment by jaesoni** [comment]
#sign-off

Please fetch and merge changes from https://github.com/Azure/azure-quickstart-templates/pull/12204 before merging this one. Thanks

**Review by mainak-agrawal — APPROVED** [review]

File: demos/ag-alert-keyvault-advisor/README.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+This template deploys an Azure Monitor alert which notifies you in an event **an Application Gateway in the given resource group is facing Key Vault related issues**.
+
+**Notes**
+
+- Azure Monitor alert rules are charged based on the type and number of signals it monitors. You may want to visit the [pricing page](https://azure.microsoft.com/en-in/pricing/details/monitor/) before deploying this alert template or can view the estimated cost after the deployment. 
+
+- You will need to create [Azure Monitor Action Group](https://docs.microsoft.com/en-us/azure/azure-monitor/alerts/action-groups) in advance and provide its ResourceID during this deployment. This action group notifies your users when the alert rule is triggered. You can use an existing or create a new one and reuse it for multiple such alerts.
+
+>[!TIP]
```
Comment (bmoore-msft) [inline demos/ag-alert-keyvault-advisor/README.md:24]: this is not valid markdown for github

File: demos/ag-alert-keyvault-advisor/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,64 @@
+{
+   "$schema":"https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+   "contentVersion":"1.0.0.0",
+   "parameters":{
+      "alertName":{
+         "type":"String"
+      },
+      "applicationGatewayResourceGroup":{
```
Comment (bmoore-msft) [inline demos/ag-alert-keyvault-advisor/azuredeploy.json:8]: change this to prompt only for the resourceGroup name (and sub if needed) and create the resourceId in the template

File: demos/ag-alert-keyvault-advisor/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+      "alertName":{
+         "type":"String"
+      },
+      "applicationGatewayResourceGroup":{
+         "type":"String",
+         "metadata":{
+            "description":"Resource Group example /subscriptions/subs-id/resourceGroups/rg-name"
+         }
+      },
+      "actionGroupResourceID":{
```
Comment (bmoore-msft) [inline demos/ag-alert-keyvault-advisor/azuredeploy.json:14]: change this to prompt for the sub, rg and resource name - you can create the resourceId for the user in the template

File: demos/ag-alert-keyvault-advisor/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+         "location":"Global",
+         "properties":{
+            "scopes":[
+               "[parameters('applicationGatewayResourceGroup')]"
+            ],
+            "condition":{
+               "allOf":[
+                  {
+                     "field":"properties.recommendationType",
+                     "equals":"3467464b-955a-4caf-95e5-547344ba0281"
```
Comment (bmoore-msft) [inline demos/ag-alert-keyvault-advisor/azuredeploy.json:38]: how does the user know what this guid indicates?

File: demos/ag-alert-keyvault-advisor/prereqs/prereq.azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+        "name": "[variables('actionGroupName')]",
+        "location": "Global",
+        "properties": {
+            "groupShortName": "[variables('actionGroupName')]",
+            "enabled": true
+        }
+    }
+  ],
+  "outputs": {
+        "applicationGatewayResourceGroup": {
```
Comment (bmoore-msft) [inline demos/ag-alert-keyvault-advisor/prereqs/prereq.azuredeploy.json:21]: you'll need to update these to match the parameter changes in the main template...

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: demos/ag-alert-keyvault-advisor/azuredeploy.json (hunk 10 above)
Comment (jaesoni) [reply demos/ag-alert-keyvault-advisor/azuredeploy.json:8]: As an MVP for these quick deployment of Azure Monitor alerts, we have taken this design decision to let customers provide the ResourceIDs manually (for both - an existing Application Gateway and Action Group). We can refine this experience in the further iterations. The same experience is used for the already published https://github.com/Azure/azure-quickstart-templates/blob/master/demos/ag-alert-backend-5xx/README.md

File: demos/ag-alert-keyvault-advisor/azuredeploy.json (hunk 11 above)
Comment (jaesoni) [reply demos/ag-alert-keyvault-advisor/azuredeploy.json:14]: https://github.com/Azure/azure-quickstart-templates/pull/12205#discussion_r812779879

File: demos/ag-alert-keyvault-advisor/azuredeploy.json (hunk 12 above)
Comment (jaesoni) [reply demos/ag-alert-keyvault-advisor/azuredeploy.json:38]: The user doesn't need to know the GUID of the alert type. The description of this alert "Setup alert for misconfigured Key Vault for your Azure Application Gateway" signifies what the alert is meant for and will be deplyed for that precise requirement.

File: demos/ag-alert-keyvault-advisor/README.md (hunk 9 above)
Comment (jaesoni) [reply demos/ag-alert-keyvault-advisor/README.md:24]: https://github.com/Azure/azure-quickstart-templates/pull/12203#discussion_r812769866

Comment (bmoore-msft) [reply demos/ag-alert-keyvault-advisor/README.md:24]: Good catch - we should fix both of them...

File: demos/ag-alert-keyvault-advisor/azuredeploy.json (hunk 12 above)
Comment (bmoore-msft) [reply demos/ag-alert-keyvault-advisor/azuredeploy.json:38]: Where do I find the list of guids/alerts?

File: demos/ag-alert-keyvault-advisor/README.md (hunk 9 above)
Comment (jaesoni) [reply demos/ag-alert-keyvault-advisor/README.md]: fixed

**Comment by azure-quickstarts** [comment]
@jaesoni - check this PR for updates that may be needed to documentation that references this sample.  [This is an automated message. You are receiving it because you are listed as the docOwner in metadata.json.]

File: demos/ag-alert-keyvault-advisor/azuredeploy.json (hunk 12 above)
Comment (jaesoni) [reply demos/ag-alert-keyvault-advisor/azuredeploy.json:38]: https://docs.microsoft.com/en-us/rest/api/advisor/recommendation-metadata/list This should help

However, related to our product there are barely 3-4 alerts, and of those we want to nudge customers for this specific critical alert only.

Comment (bmoore-msft) [reply demos/ag-alert-keyvault-advisor/azuredeploy.json:38]: Is there something we can put into the readme so customers can learn more?

Comment (jaesoni) [reply demos/ag-alert-keyvault-advisor/azuredeploy.json:38]: That is outside of scope of Application Gateway product. We have only one alert that is focusing on our product's scenario

**Pull request merged**